### PR TITLE
Implement Trusted Vector Preallocation

### LINKED ISSUE #1880: Safe Preallocation for (most) `Vec<T>`
# Overview - Safe Preallocation
Currently, Zebra [doesn't preallocate Vectors](https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/serialization/zcash_deserialize.rs) when deserializing, since blind preallocations present a DOS vector. Instead, it relies on maximum block and transaction size to limit allocations. This is inefficient.

We can mitigate the DOS potential and allow for preallocation as follows:

1. Define a new trait `SafeAllocate`, defining the maximum length a Vec<T: SafeAllocate> can sensibly reach for each implementing type
1. Create a specialized `ZcashDeserialize` implementation for `SafeAllocate`rs.

```
pub trait SafeAllocate {
    const fn max_allocation() -> usize;
}

impl<T: ZcashDeserialize + SafeAllocate> ZcashDeserialize for Vec<T> {
    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
        let len = reader.read_compactsize()?;
        if len > T::max_allocation() {
            return Err(SerializationError::Parse(("Vector longer than max_allocation")));
        }
        let mut vec = Vec::with_capacity(len);
        for _ in 0..len {
            vec.push(T::zcash_deserialize(&mut reader)?);
        }
        Ok(vec)
    }
}
```

This allows us to pre-allocate for certain types without presenting a DOS vector, while retaining the flexibilty to use the current (lazy) allocation strategy for types that defy safe blind allocation.

Note that deserialization using this method is guaranteed to fail during the deserialization of the first malicious vector. However, Block messages contain nested `Vec<T>` types, and it is possible for a Byzantine message to cause improper allocations at each level of nesting before being discovered. This is why the potential spike for a Block message is much higher than for other message types.

# Potential SafeAllocation Implementors

- `Arc<Transaction>` - Suggested allocation limit: 49_000. A Zcash block is capped at 2MB. Since transparent Txs are the smallest allowable transactions, and each transparent Tx requires at least one Input with a minimum serialized size of 41 bytes, an upper bound the length of a `Vec<Arc<Transaction>>` is 2MB / 41 Bytes < 49_000. Note that `std::mem::size_of::<Arc<T>>()` is 8, so the maximum memory wasted by a Byzantine `Vec<Arc<Transaction>>` is 49_000 \* 8 Bytes < 400KB.

- `transparent::Input` - Suggested allocation limit: 49_000. A Zcash block is capped at 2MB. Each Input has a minimum serialized size of 41 bytes, so an upper bound the length of a `Vec<transparent::Input>` is 2MB / 41 Bytes < 49_000. With Inputs taking less than 80 bytes on the stack (36 for the Outpoint, 26 for the Script, and 4 for the sequence), the maximum memory wasted by a Byzantine `Vec<transparent::Input>` is 49_000 \* 80 Bytes < 4MB.

- `transparent::Output` - Suggested allocation limit: 225_000. A Zcash block is capped at 2MB. Outputs have a minimum serialized size of 9 bytes, so an upper bound the length of a `Vec<Arc<transparent::Output>>` is 2MB / 9 Bytes < 225_000. With Outputs taking less than 40 bytes on the stack (call it 10 for the value, 26 for the Script, for a total of 36 ), the maximum memory wasted by a Byzantine `Vec<Arc<transparent::Output>>` is 225_000 \* 40 Bytes = 9MB.

- `MetaAddr` - Suggested allocation  limit: 1000. No fancy math required, since this limit is in the Addr message [specification](https://developer.bitcoin.org/reference/p2p_networking.html#addr). Estimate a MetaAddr at a generous 100 bytes of stack space and you get max memory waste = 1000 \* 100B = 100KB.

- `block::Hash` - Suggested allocation limit: 65536. We derive this limit as MAX_MESSAGE_SIZE / 32 = 2 \* 1024 \_ 1024 / 32. Since a block hash takes 32 bytes on the stack, the max waste here is MAX_MESSAGE_SIZE.

- `InventoryHash` - Suggested allocation  limit: 50,000. This limit is the listed in the [Inv message spec](https://developer.bitcoin.org/reference/p2p_networking.html#inv). Maximum waste: 50,000 \* 32 B = 1.6 MB.

- `block::CountedHeader` - Suggested allocation limit: 2000. This limit is in the [Headers message spec](https://developer.bitcoin.org/reference/p2p_networking.html#headers). Per the most recent spec, each [Zcash header is less than 2kb](https://zips.z.cash/protocol/protocol.pdf#page=90&zoom=100,72,73). Maximum waste: 2000 \* 2000 Bytes < 4MB

- `u8` - Suggested allocation limit: MAX_MESSAGE_SIZE. Since a u8 takes 1 byte on the stack, the maximum waste here is MAX_MESSAGE_SIZE = 2MB.

## Example attacks

Using all of these numbers, we calculate the total improper allocation caused by the worst-case instance of each Zcash Message as follows:

- Version: N.A.
- Verack: N.A.
- Ping: N.A.
- Pong: N.A.
- Reject: N.A.
- GetAddr: N.A.
- Addr: contains 1 `Vec<MetaAddr>` => **100KB** (see above)
- Get locks: contains 1 `Vec<block::Hash>` => **MAX_MESSAGE_SIZE** (2MB)
- Inv: contains 1 `Vec<InventoryHash>` => **1.6 MB** (see above)
- GetHeaders: contains 1 `Vec<block::Hash>` => **MAX_MESSAGE_SIZE** (2MB)
- Headers: contains 1 `Vec<block::CountedHeader>` => **4MB**
- GetData: contains 1 `Vec<InventoryHash>` => **1.6 MB** (see above)
- Block: The worst case block contains 1 `Vec<Arc<Transaction>`, 1 `Vec<transparent::Output>`, and 1 malicious `Script(Vec<u8>)` => 400KB + 9MB + 2MB = **11.4 MB**.
- - Note that a dishonest vector is discovered during its own deserialization, so a malicious `Vec<transparent::Input>` would be discovered before the malicious `Vec<transparent::Output>` was allocated for. Since Outputs can waste more memory than Inputs, a smart attacker will choose to make only his `Vec<transparent::Output>` malicious.
- Tx: N.A.
- NotFound: contains 1 `Vec<InventoryHash>` => **1.6 MB** (see above)
- Mempool: N.A.
- FilterLoad: contains 1 `Filter(Vec<u8>)` => **2MB**
- FilterAdd: contains 1 `Vec<u8>` => **2MB**
- FilterClear: N.A.

# Alternatives
Do nothing. This is a fine option, since Zebra's networking stack is already fast! 

# Summary and Recommendations

With the `SafeAllocate` trait, we can allow preallocation for many Vector types without risking DOS attacks.

In the worst case, a malicious message can cause a short-lived spike in memory usage. The size of this spike depends on the max_allocation defined in `SafeAllocate` and the depth of nested `Vec<T: SafeAllocate>` types. Calculations of the maximum spike caused by each message type are included above. Based on these calculations, I recommend implementing SafeAllocate for all types listed in the "Potential SafeAllocate Implementors" section, with the possible exception of `transparent::Input` and `transparent::Output`.

If this recommendation is adopted , the worst case memory spike that a malicious peer can cause will be roughly 4MB, or roughly double that peer connection's usual memory consumption.

If `transparent::Input`s and `transparent::Output`s are included, the worst case spike rises to 11.5MB, or about six times a peer connection's usual memory consumption.

### PR DESCRIPTION
**Pull request by preston-evans98**
## Motivation
Resolve #1880.

## Solution
Implement the `SafePreallocate` trait, setting sane limits on blind preallocation of vectors. Implement proptests (where applicable) and unit tests to verify that the limits set by SafePreallocate are loose enough that all structurally valid messages can be deserialized. 

This implementation deviates from the specification in #1880  in order to remove the use  of `min_specialization`. Instead, we tighten the trait bound from  `impl<T> zcash_deserialize for Vec<T>`, to  `impl<T: SafePreallocate> zcash_deserialize for Vec<T>`, and implement `SafePreallocate` for all necessary types. If we need to deserialize a Vector type which can't be safely PreAllocated in future, we'll just have to write a manual impl for it. 

Also note that the bounds in this implementation differ slightly from those in #1880 in order to remove special cases. For example, we don't set the `max_allocation()` for `MetaAddr` to 1000 as described in the Bitcoin specification. Instead, we set the limit based on the `MAX_PROTOCOL_MESSAGE_SIZE` and the serialized length of a MetaAddr.  This mitigates the risk of that future changes to the Zcash protocol will render the bounds in SafePreallocate obsolete, potentially causing Zebra to fall out of consensus. 

The code in this pull request has:
  - [x]  Documentation Comments
  - [x] Unit Tests and Property Tests

## Review
This is low priority. @teor2345 

## Related Issues
#1880, #1917

## Follow Up Work
Improve the efficiency of the proptests for `SafePreallocate`.

### CODE REVIEW

File: zebra-chain/src/block.rs [hunk 1]
```diff
@@ -80,3 +87,38 @@ impl<'a> From<&'a Block> for Hash {
...
+            let mut smallest_disallowed_vec = Vec::with_capacity(max_allocation + 1);
+            for _ in 0..(Hash::max_allocation()+1) {
+                smallest_disallowed_vec.push(hash.clone());
+            }
+            let serialized = smallest_disallowed_vec.zcash_serialize_to_vec().expect("Serialization to vec must succeed");
+
+            // Check that our smallest_disallowed_vec is only one item larger than the limit
+            prop_assert!(((smallest_disallowed_vec.len() - 1) as u64) == Hash::max_allocation());
+            // Check that our smallest_disallowed_vec is too big to send as a protocol message
+            prop_assert!(serialized.len() >= MAX_PROTOCOL_MESSAGE_LEN);
```
Comment (teor2345) [inline zebra-chain/src/block.rs]: Nitpick:
```suggestion
            // Check that our smallest_disallowed_vec is too big to send as a protocol message
            // Since every current message type has a size field, the data can't be equal to the maximum size
            prop_assert!(serialized.len() >= MAX_PROTOCOL_MESSAGE_LEN);
```

File: zebra-chain/src/block.rs [hunk 2]
```diff
@@ -80,3 +87,38 @@ impl<'a> From<&'a Block> for Hash {
...
+
+
+        #[test]
+        fn block_hash_max_allocation(hash in Hash::arbitrary_with(())) {
+            let max_allocation: usize = Hash::max_allocation().try_into().unwrap();
+            let mut smallest_disallowed_vec = Vec::with_capacity(max_allocation + 1);
+            for _ in 0..(Hash::max_allocation()+1) {
+                smallest_disallowed_vec.push(hash.clone());
+            }
+            let serialized = smallest_disallowed_vec.zcash_serialize_to_vec().expect("Serialization to vec must succeed");
```
Comment (teor2345) [inline zebra-chain/src/block.rs]: Can we also test that the size of the `largest_allowed_vec` is less than the `MAX_PROTOCOL_MESSAGE_LEN` in each proptest?

File: zebra-chain/src/block/header.rs [hunk 3]
```diff
@@ -118,3 +123,59 @@ pub struct CountedHeader {
     pub header: Header,
     pub transaction_count: usize,
 }
+
+// Includes the 32-byte nonce and 3-byte equihash length field.
```
Comment (teor2345) [inline zebra-chain/src/block/header.rs]: Might as well make this a doc comment.
```suggestion
/// The serialized size of a Zcash block header.
///
/// Includes the equihash input, 32-byte nonce, 3-byte equihash length field, and equihash solution.
```

File: zebra-chain/src/block/header.rs [hunk 4]
```diff
@@ -118,3 +123,59 @@ pub struct CountedHeader {
...
+const BLOCK_HEADER_LENGTH: usize =
+    crate::work::equihash::Solution::INPUT_LENGTH + 32 + 3 + crate::work::equihash::SOLUTION_SIZE;
+
+const COUNTED_HEADER_LEN: usize = BLOCK_HEADER_LENGTH + 1;
+impl SafePreallocate for CountedHeader {
+    fn max_allocation() -> u64 {
+        // A CountedHeader has BLOCK_HEADER_LENGTH bytes + 1 byte for the transaction count
+        (MAX_PROTOCOL_MESSAGE_LEN / COUNTED_HEADER_LEN) as u64
```
Comment (teor2345) [inline zebra-chain/src/block/header.rs]: Nitpick: the transaction count is a `var_int`, so it's the minimum size.

Might as well make this a doc comment.
```suggestion
/// The minimum size for a serialized CountedHeader.
///
/// A CountedHeader has BLOCK_HEADER_LENGTH bytes + 1 or more bytes for the transaction count
const MIN_COUNTED_HEADER_LEN: usize = BLOCK_HEADER_LENGTH + 1;
impl SafePreallocate for CountedHeader {
    fn max_allocation() -> u64 {
        (MAX_PROTOCOL_MESSAGE_LEN / MIN_COUNTED_HEADER_LEN) as u64
```

File: zebra-chain/src/block.rs [hunk 5]
```diff
@@ -80,3 +87,38 @@ impl<'a> From<&'a Block> for Hash {
...
+mod test_safe_preallocate {
+    use super::{Hash, MAX_PROTOCOL_MESSAGE_LEN};
+    use crate::serialization::{SafePreallocate, ZcashSerialize};
+    use proptest::prelude::*;
+    use std::convert::TryInto;
+    proptest! {
+
+        #![proptest_config(ProptestConfig::with_cases(200))]
+
+
```
Comment (teor2345) [inline zebra-chain/src/block.rs]: Can we have a short doc comment explaining the purpose of each test?

File: zebra-chain/src/block/header.rs [hunk 6]
```diff
@@ -118,3 +123,59 @@ pub struct CountedHeader {
...
+        fn counted_header_min_length(header in Header::arbitrary_with(()), transaction_count in (0..std::u32::MAX)) {
+            let header = CountedHeader {
+                header,
+                transaction_count: transaction_count.try_into().expect("Must run test on platform with at least 32 bit address space"),
+            };
+            let serialized_header = header.zcash_serialize_to_vec().expect("Serialization to vec must succeed");
+            prop_assert!(serialized_header.len() >= COUNTED_HEADER_LEN)
+        }
+
+
```
Comment (teor2345) [inline zebra-chain/src/block/header.rs]: TODO: test doc comment and `largest_allowed_vec`.

File: zebra-chain/src/sapling/output.rs [hunk 7]
```diff
@@ -51,3 +54,53 @@ impl ZcashDeserialize for Output {
...
+mod test_safe_preallocate {
+    use super::{Output, MAX_BLOCK_BYTES, OUTPUT_SIZE};
+    use crate::serialization::{SafePreallocate, ZcashSerialize};
+    use proptest::prelude::*;
+    use std::convert::TryInto;
+
+    proptest! {
+        #![proptest_config(ProptestConfig::with_cases(10_000))]
+
+        /// Confirm that each output takes at least OUTPUT_SIZE bytes when serialized.
```
Comment (teor2345) [inline zebra-chain/src/sapling/output.rs]:
```suggestion
        /// Confirm that each output takes exactly OUTPUT_SIZE bytes when serialized.
```

File: zebra-chain/src/sapling/output.rs [hunk 8]
```diff
@@ -51,3 +54,53 @@ impl ZcashDeserialize for Output {
...
+        #[test]
+        fn output_size_is_small_enough(output in Output::arbitrary_with(())) {
+            let serialized = output.zcash_serialize_to_vec().expect("Serialization to vec must succeed");
+            prop_assert!(serialized.len() as u64 == OUTPUT_SIZE)
+        }
+
+    }
+    proptest! {
+        #![proptest_config(ProptestConfig::with_cases(100))]
+        #[test]
```
Comment (teor2345) [inline zebra-chain/src/sapling/output.rs:123]: TODO: test doc comment and `largest_allowed_vec`.

File: zebra-chain/src/sapling/output.rs [hunk 9]
```diff
@@ -51,3 +54,53 @@ impl ZcashDeserialize for Output {
         })
     }
 }
+/// An output contains: a 32 byte cv, a 32 byte cmu, a 32 byte ephemeral key
+/// a 580 byte encCiphertext, an 80 byte outCiphertext, and a 192 byte zkproof
+/// [ps]: https://zips.z.cash/protocol/protocol.pdf#outputencoding
+const OUTPUT_SIZE: u64 = 32 + 32 + 32 + 580 + 80 + 192;
```
Comment (teor2345) [inline zebra-chain/src/sapling/output.rs:84]: The Sapling changes might conflict with some NU5/Orchard work we're doing right now, but we'll handle that if it comes up.

FYI @oxarbitrage @dconnolly

File: zebra-chain/src/sapling/spend.rs [hunk 10]
```diff
@@ -56,3 +58,54 @@ impl ZcashDeserialize for Spend {
...
+mod test_safe_preallocate {
+    use super::{Spend, MAX_BLOCK_BYTES, SPEND_SIZE};
+    use crate::serialization::{SafePreallocate, ZcashSerialize};
+    use proptest::prelude::*;
+    use std::convert::TryInto;
+
+    proptest! {
+        #![proptest_config(ProptestConfig::with_cases(10_000))]
+
+        /// Confirm that each spend takes at least SPEND_SIZE bytes when serialized.
```
Comment (teor2345) [inline zebra-chain/src/sapling/spend.rs]:
```suggestion
        /// Confirm that each spend takes exactly SPEND_SIZE bytes when serialized.
```

File: zebra-chain/src/sapling/spend.rs [hunk 11]
```diff
@@ -56,3 +58,54 @@ impl ZcashDeserialize for Spend {
...
+        #[test]
+        fn spend_size_is_small_enough(spend in Spend::arbitrary_with(())) {
+            let serialized = spend.zcash_serialize_to_vec().expect("Serialization to vec must succeed");
+            prop_assert!(serialized.len() as u64 == SPEND_SIZE)
+        }
+
+    }
+    proptest! {
+        #![proptest_config(ProptestConfig::with_cases(100))]
+        #[test]
```
Comment (teor2345) [inline zebra-chain/src/sapling/spend.rs]: TODO: test doc comment and `largest_allowed_vec`.

File: zebra-chain/src/serialization/error.rs [hunk 12]
```diff
@@ -13,6 +13,9 @@ pub enum SerializationError {
     // XXX refine errors
     #[error("parse error: {0}")]
     Parse(&'static str),
+    /// The length of a vec is too large to convert to a usize (and thus, too large to allocate on this platform)
```
Comment (teor2345) [inline zebra-chain/src/serialization/error.rs:16]: 👍

File: zebra-chain/src/serialization/proptests.rs [hunk 13]
```diff
@@ -6,6 +6,7 @@ use proptest::prelude::*;
 use std::io::Cursor;
 
 proptest! {
+
```
Comment (teor2345) [inline zebra-chain/src/serialization/proptests.rs]: Nitpick: we try to avoid unnecessary whitespace changes
```suggestion
```

File: zebra-chain/src/serialization/zcash_deserialize.rs [hunk 14]
```diff
@@ -49,3 +49,15 @@ impl<R: io::Read> ZcashDeserializeInto for R {
...
 }
+
+/// Blind preallocation of a Vec<T: SafePreallocate> can be done safely. This is in contrast
+/// to blind preallocation of a generic Vec<T>, which is a DOS vector.
+///
+/// The max_allocation() function provides a loose upper bound on the size of the Vec<T: SafePreallocate>
+/// which can possibly be received from an honest peer. If this limit is too low, Zebra may reject valid messages.
+/// In the worst case, setting the lower bound to low could cause Zebra to fall out of consensus by rejecting all messages containing a valid block.
+pub trait SafePreallocate {
+    /// Provides a ***loose upper bound*** on the size of the Vec<T: SafePreallocate>
```
Comment (teor2345) [inline zebra-chain/src/serialization/zcash_deserialize.rs]: 👍

File: zebra-chain/src/transaction/serialize.rs [hunk 15]
```diff
@@ -342,3 +343,119 @@ where
...
+
+/// A Tx Input must have an Outpoint (32 byte hash + 4 byte index), a 4 byte sequence number,
+/// and a signature script, which always takes a min of 1 byte (for a length 0 script)
+const MIN_TRANSPARENT_INPUT_SIZE: u64 = 32 + 4 + 4 + 1;
+/// A Transparent output has an 8 byte value and script which takes a min of 1 byte
+const MIN_TRANSPARENT_OUTPUT_SIZE: u64 = 8 + 1;
+// All txs must have at least one input and a 4 byte locktime
+const MIN_TRANSPARENT_TX_SIZE: u64 = MIN_TRANSPARENT_INPUT_SIZE + 4;
+
+/// No valid Zcash message contains more transactions than can fit in a single block
```
Comment (teor2345) [inline zebra-chain/src/transaction/serialize.rs:366]:
```suggestion
/// No valid Zcash message contains more transactions than can fit in a single block
///
/// `tx` messages contain a single transaction, and `block` messages are limited to the maximum
/// block size.
```

File: zebra-chain/src/transaction/serialize.rs [hunk 16]
```diff
@@ -342,3 +343,119 @@ where
...
+impl SafePreallocate for Arc<Transaction> {
+    fn max_allocation() -> u64 {
+        MAX_BLOCK_BYTES / MIN_TRANSPARENT_TX_SIZE
+    }
+}
+/// No valid Zcash message contains more transactions inputs than can fit in maximally large transaction
```
Comment (teor2345) [inline zebra-chain/src/transaction/serialize.rs]:
```suggestion
/// The maximum number of inputs in a valid Zcash on-chain transaction.
///
/// If a transaction contains more inputs than can fit in maximally large block, it might be
/// valid on the network and in the mempool, but it can never be mined into a block. So
/// rejecting these large edge-case transactions can never break consensus.
```

File: zebra-chain/src/transaction/serialize.rs [hunk 17]
```diff
@@ -342,3 +343,119 @@ where
...
+impl SafePreallocate for transparent::Input {
+    fn max_allocation() -> u64 {
+        MAX_BLOCK_BYTES / MIN_TRANSPARENT_INPUT_SIZE
+    }
+}
+/// No valid Zcash message contains more transactions outputs than can fit in maximally large transaction
```
Comment (teor2345) [inline zebra-chain/src/transaction/serialize.rs]:
```suggestion
/// The maximum number of outputs in a valid Zcash on-chain transaction.
///
/// If a transaction contains more outputs than can fit in maximally large block, it might be
/// valid on the network and in the mempool, but it can never be mined into a block. So
/// rejecting these large edge-case transactions can never break consensus.
```

File: zebra-chain/src/transaction/serialize.rs [hunk 18]
```diff
@@ -342,3 +343,119 @@ where
...
+        /// This verifies that our calculated `SafePreallocate::max_allocation()` is indeed an upper bound.
+        #[test]
+        fn transparent_output_size_is_small_enough(output in Output::arbitrary()) {
+            let serialized = output.zcash_serialize_to_vec().expect("Serialization to vec must succeed");
+            prop_assert!(serialized.len() as u64 >= MIN_TRANSPARENT_OUTPUT_SIZE)
+        }
+
+    }
+    proptest! {
+        // This test is pretty slow, so only run a few
```
Comment (teor2345) [inline zebra-chain/src/transaction/serialize.rs:435]: We try to make sure each individual test runs for 10 seconds or less, but we haven't checked test speeds for a while

File: zebra-chain/src/transaction/serialize.rs [hunk 19]
```diff
@@ -342,3 +343,119 @@ where
...
+        /// Confirm that each spend takes at least MIN_TRANSPARENT_TX_SIZE bytes when serialized.
+        /// This verifies that our calculated `SafePreallocate::max_allocation()` is indeed an upper bound.
+        #[test]
+        fn transparent_output_size_is_small_enough(output in Output::arbitrary()) {
+            let serialized = output.zcash_serialize_to_vec().expect("Serialization to vec must succeed");
+            prop_assert!(serialized.len() as u64 >= MIN_TRANSPARENT_OUTPUT_SIZE)
+        }
+
+    }
+    proptest! {
```
Comment (teor2345) [inline zebra-chain/src/transaction/serialize.rs:434]: TODO: test doc comment and `largest_allowed_vec`.

File: zebra-network/src/meta_addr.rs [hunk 20]
```diff
@@ -200,6 +201,58 @@ impl ZcashDeserialize for MetaAddr {
...
+
+#[cfg(test)]
+mod test_safe_preallocate {
+    use std::convert::TryInto;
+
+    use super::{MetaAddr, MAX_PROTOCOL_MESSAGE_LEN, META_ADDR_SIZE};
+    use super::{PeerAddrState, PeerServices};
+    use chrono::{TimeZone, Utc};
+    use zebra_chain::serialization::{SafePreallocate, ZcashSerialize};
+    #[test]
```
Comment (teor2345) [inline zebra-network/src/meta_addr.rs:285]: TODO: test doc comment and `largest_allowed_vec`.

File: zebra-chain/src/block.rs [hunk 21]
```diff
@@ -80,3 +87,38 @@ impl<'a> From<&'a Block> for Hash {
...
+    }
+}
+
+#[cfg(test)]
+mod test_safe_preallocate {
+    use super::{Hash, MAX_PROTOCOL_MESSAGE_LEN};
+    use crate::serialization::{SafePreallocate, ZcashSerialize};
+    use proptest::prelude::*;
+    use std::convert::TryInto;
+    proptest! {
```
Comment (teor2345) [inline zebra-chain/src/block.rs:107]: Nitpick: Can we have a serialized size test for `Hash`, for completeness?

File: zebra-network/src/protocol/external/inv.rs [hunk 22]
```diff
@@ -81,3 +85,10 @@ impl ZcashDeserialize for InventoryHash {
         }
     }
 }
+
+impl SafePreallocate for InventoryHash {
```
Comment (teor2345) [inline zebra-network/src/protocol/external/inv.rs]: Nitpick: Can we have the standard set of tests for `InventoryHash`, for completeness?

**Review by teor2345 — CHANGES_REQUESTED** [review]
Thanks for doing all this work, it looks great.

I'd like some tests to make sure that the maximum preallocation is within the maximum block or message size. Otherwise, the code is really good.

I also have a bunch of suggestions and nitpicks around documentation.

While reviewing this PR, I discovered that Zebra preallocates a few other types. So after the fix in #1925 merges, this PR will need a `SafePreallocate` impl for `JoinSplitData`, `String`, and `Script`.

File: zebra-chain/src/block.rs [hunk 23]
```diff
@@ -80,3 +87,38 @@ impl<'a> From<&'a Block> for Hash {
...
+
+        #![proptest_config(ProptestConfig::with_cases(200))]
+
+
+        #[test]
+        fn block_hash_max_allocation(hash in Hash::arbitrary_with(())) {
+            let max_allocation: usize = Hash::max_allocation().try_into().unwrap();
+            let mut smallest_disallowed_vec = Vec::with_capacity(max_allocation + 1);
+            for _ in 0..(Hash::max_allocation()+1) {
+                smallest_disallowed_vec.push(hash.clone());
```
Comment (teor2345) [inline zebra-chain/src/block.rs]: Clippy: `clone` isn't needed, because `hash` implements `Copy`
```suggestion
                smallest_disallowed_vec.push(hash);
```

File: zebra-chain/src/transaction/serialize.rs [hunk 24]
```diff
@@ -342,3 +343,119 @@ where
...
     }
 }
+
+/// A Tx Input must have an Outpoint (32 byte hash + 4 byte index), a 4 byte sequence number,
+/// and a signature script, which always takes a min of 1 byte (for a length 0 script)
+const MIN_TRANSPARENT_INPUT_SIZE: u64 = 32 + 4 + 4 + 1;
+/// A Transparent output has an 8 byte value and script which takes a min of 1 byte
+const MIN_TRANSPARENT_OUTPUT_SIZE: u64 = 8 + 1;
+// All txs must have at least one input and a 4 byte locktime
+const MIN_TRANSPARENT_TX_SIZE: u64 = MIN_TRANSPARENT_INPUT_SIZE + 4;
```
Comment (teor2345) [inline zebra-chain/src/transaction/serialize.rs]: The consensus rules in Zcash are slightly different, each transparent transaction must have an input *and* an output.

Screen Shot 2021-03-18 at 08 41 45

File: zebra-chain/src/transaction/serialize.rs [hunk 25]
```diff
@@ -342,3 +343,119 @@ where
...
+
+/// No valid Zcash message contains more transactions than can fit in a single block
+impl SafePreallocate for Arc<Transaction> {
+    fn max_allocation() -> u64 {
+        MAX_BLOCK_BYTES / MIN_TRANSPARENT_TX_SIZE
```
Comment (teor2345) [inline zebra-chain/src/transaction/serialize.rs:373]:
```suggestion
        // A transparent transaction is the smallest transaction variant
        MAX_BLOCK_BYTES / MIN_TRANSPARENT_TX_SIZE
```

**Review by teor2345 — COMMENTED** [review]
I also discovered `String` and `Script` preallocations, which I fixed in #1925.

Note that `equihash::Solution` and `transparent::Input::Coinbase` also do manual allocations, but their checks are very strict.

File: zebra-chain/src/block.rs (hunk 2 above)
Comment (preston-evans98) [reply zebra-chain/src/block.rs]: We can for some types (Hash, Spend, Output, MetAddr, InventoryHash, and Header) but not for the types with variable sizes (all of the Tx types). I'll go ahead and add that check for all of the fixed-size types.

File: zebra-chain/src/block.rs (hunk 1 above)
Comment (preston-evans98) [reply zebra-chain/src/block.rs]: On second thought, I think my original implementation is off by one. I delegate serialization to `zcash_deserialize_to_vec()` which does include the length field. I'll switch this to a strict inequality in all of the tests.

File: zebra-chain/src/serialization/zcash_deserialize.rs [hunk 26]
```diff
@@ -49,3 +49,15 @@ impl<R: io::Read> ZcashDeserializeInto for R {
...
     }
 }
+
+/// Blind preallocation of a Vec<T: SafePreallocate> can be done safely. This is in contrast
+/// to blind preallocation of a generic Vec<T>, which is a DOS vector.
+///
+/// The max_allocation() function provides a loose upper bound on the size of the Vec<T: SafePreallocate>
+/// which can possibly be received from an honest peer. If this limit is too low, Zebra may reject valid messages.
+/// In the worst case, setting the lower bound to low could cause Zebra to fall out of consensus by rejecting all messages containing a valid block.
+pub trait SafePreallocate {
```
Comment (teor2345) [inline zebra-chain/src/serialization/zcash_deserialize.rs]: `unsafe` is a Rust keyword with a very specific meaning:
https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html

Memory corruption is covered by Rust's safety guarantees, but memory denial of service isn't. So let's use "trusted" instead of "safe", to avoid confusion.
```suggestion
/// Blind preallocation of a Vec<T: TrustedPreallocate> is based on a bounded length. This is in contrast
/// to blind preallocation of a generic Vec<T>, which is a DOS vector.
///
/// The max_allocation() function provides a loose upper bound on the size of the Vec<T: SafePreallocate>
/// which can possibly be received from an honest peer. If this limit is too low, Zebra may reject valid messages.
/// In the worst case, setting the lower bound to low could cause Zebra to fall out of consensus by rejecting all messages containing a valid block.
pub trait TrustedPreallocate {
```

File: zebra-chain/src/block.rs (hunk 5 above)
Comment (preston-evans98) [reply zebra-chain/src/block.rs]: Added

File: zebra-chain/src/block/header.rs (hunk 6 above)
Comment (preston-evans98) [reply zebra-chain/src/block/header.rs]: Added

File: zebra-chain/src/transaction/serialize.rs (hunk 17 above)
Comment (preston-evans98) [reply zebra-chain/src/transaction/serialize.rs]: Fixed in follow up commit

File: zebra-chain/src/transaction/serialize.rs (hunk 19 above)
Comment (preston-evans98) [reply zebra-chain/src/transaction/serialize.rs:434]: Done

File: zebra-network/src/protocol/external/inv.rs (hunk 22 above)
Comment (preston-evans98) [reply zebra-network/src/protocol/external/inv.rs]: Added in follow up

**Comment by teor2345** [comment]
Looks like you're ready for another review here.

I should get to it next week, but we're prioritising the Orchard network upgrade right now, so it might slip to early the week after.

**Comment by teor2345** [comment]
Looks good so far, see preston-evans98/zebra#1 for a fix to the merge conflict with #1946.

We'll probably want to move the tests to their own modules inside each test submodule. But we can do that in a follow-up PR.

I'm just doing some testing now, then I'll do a detailed review early next week.

**Comment by teor2345** [comment]
The Google Cloud Build error is a network failure downloading the Rust docker image.

File: zebra-network/src/meta_addr.rs [hunk 27]
```diff
@@ -263,6 +264,76 @@ impl ZcashDeserialize for MetaAddr {
...
+    use std::convert::TryInto;
+
+    use super::{MetaAddr, MAX_PROTOCOL_MESSAGE_LEN, META_ADDR_SIZE};
+    use super::{PeerAddrState, PeerServices};
+    use chrono::{TimeZone, Utc};
+    use zebra_chain::serialization::{TrustedPreallocate, ZcashSerialize};
+    #[test]
+    /// Confirm that each MetaAddr takes exactly META_ADDR_SIZE bytes when serialized.
+    /// This verifies that our calculated `TrustedPreallocate::max_allocation()` is indeed an upper bound.
+    fn meta_addr_size_is_correct() {
```
Comment (teor2345) [inline zebra-network/src/meta_addr.rs:288]: The `MetaAddr` tests should be proptests, but I can make that change in a follow-up PR.

File: zebra-network/src/protocol/external/inv.rs [hunk 28]
```diff
@@ -81,3 +85,85 @@ impl ZcashDeserialize for InventoryHash {
...
+    use super::{InventoryHash, INV_HASH_SIZE, MAX_PROTOCOL_MESSAGE_LEN};
+    use zebra_chain::{
+        block,
+        serialization::{TrustedPreallocate, ZcashSerialize},
+        transaction,
+    };
+    #[test]
+    /// Confirm that each InventoryHash takes exactly INV_HASH_SIZE bytes when serialized.
+    /// This verifies that our calculated `TrustedPreallocate::max_allocation()` is indeed an upper bound.
+    fn inv_hash_size_is_correct() {
```
Comment (teor2345) [inline zebra-network/src/protocol/external/inv.rs:111]: Similarly, the inv tests should be proptests, in case we add more variants in future.

File: zebra-network/src/meta_addr.rs (hunk 27 above)
Comment (teor2345) [reply zebra-network/src/meta_addr.rs:288]: (In case we add more address types in future, for example, Tor onion addresses.)

**Review by teor2345 — APPROVED** [review]
Looks good!

Let's merge this now, then I'll submit 2 small cleanup PRs.
